File: mica-webapp/src/main/webapp/app/access-config/data-access-config-service.js
/*
 * Copyright (c) 2018 OBiBa. All rights reserved.
 *
 * This program and the accompanying materials
 * are made available under the terms of the GNU Public License v3.0.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

'use strict';

mica.dataAccessConfig

  .factory('DataAccessConfigResource', ['$resource',
    function ($resource) {
      return $resource(contextPath + '/ws/config/data-access', {}, {
        'get': {method: 'GET', errorHandler: true},
        'save': {method: 'PUT', errorHandler: true}
      });
    }])

  .factory('DataAccessFormResource', ['$resource',
    function ($resource) {
      return $resource(contextPath + '/ws/config/data-access-form', {}, {
        'get': {method: 'GET', params: { revision: '@revision' }, errorHandler: true},
        'save': {method: 'PUT', errorHandler: true}
      });
    }])

  .factory('DataAccessFormPublicationResource', ['$resource',
    function ($resource) {
      return $resource(contextPath + '/ws/config/data-access-form/_publish', {}, {
        'publish': {method: 'PUT', errorHandler: true}
      });
    }])

  .factory('DataAccessFormPermissionsResource', ['$resource',
    function ($resource) {
      return $resource(contextPath + '/ws/config/data-access-form/permissions', {}, {
        'save': {
          method: 'PUT',
          params: {type: '@type', principal: '@principal', role: '@role', otherResources: '@otherResources'},
          errorHandler: true
        },
        'delete': {method: 'DELETE', params: {type: '@type', principal: '@principal'}, errorHandler: true},
        'get': {method: 'GET', isArray: true}
      });
    }])

  .factory('DataAccessFeasibilityFormResource', ['$resource',
    function ($resource) {
      return $resource(contextPath + '/ws/config/data-access-feasibility-form', {}, {
        'get': {method: 'GET', params: { revision: '@revision' }, errorHandler: true},
        'save': {method: 'PUT', errorHandler: true}
      });
    }])

  .factory('DataAccessFeasibilityFormPublicationResource', ['$resource',
    function ($resource) {
      return $resource(contextPath + '/ws/config/data-access-feasibility-form/_publish', {}, {
        'publish': {method: 'PUT', errorHandler: true}
      });
    }])

  .factory('DataAccessAgreementFormResource', ['$resource',
    function ($resource) {
      return $resource(contextPath + '/ws/config/data-access-agreement-form', {}, {
        'get': {method: 'GET', params: { revision: '@revision' }, errorHandler: true},
        'save': {method: 'PUT', errorHandler: true}
      });
    }])

  .factory('DataAccessAgreementFormPublicationResource', ['$resource',
    function ($resource) {
      return $resource(contextPath + '/ws/config/data-access-agreement-form/_publish', {}, {
        'publish': {method: 'PUT', errorHandler: true}
      });
    }])

  .factory('DataAccessAmendmentFormResource', ['$resource',
    function ($resource) {
      return $resource(contextPath + '/ws/config/data-access-amendment-form', {}, {
        'get': {method: 'GET', params: { revision: '@revision' }, errorHandler: true},
        'save': {method: 'PUT', errorHandler: true}
      });
    }])

  .factory('DataAccessAmendmentFormPublicationResource', ['$resource',
    function ($resource) {
      return $resource(contextPath + '/ws/config/data-access-amendment-form/_publish', {}, {
        'publish': {method: 'PUT', errorHandler: true}
      });
    }]);

File: mica-core/src/main/java/org/obiba/mica/micaConfig/service/DataAccessConfigService.java
/*
 * Copyright (c) 2022 OBiBa. All rights reserved.
 *
 * This program and the accompanying materials
 * are made available under the terms of the GNU Public License v3.0.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.obiba.mica.micaConfig.service;

import com.google.common.base.Strings;
import com.google.common.eventbus.EventBus;
import org.obiba.mica.micaConfig.domain.DataAccessConfig;
import org.obiba.mica.micaConfig.event.DataAccessConfigUpdatedEvent;
import org.obiba.mica.micaConfig.repository.DataAccessConfigRepository;
import org.springframework.beans.BeanUtils;
import org.springframework.stereotype.Service;
import org.springframework.util.StringUtils;
import org.springframework.validation.annotation.Validated;

import jakarta.inject.Inject;
import javax.validation.Valid;
import javax.validation.constraints.NotNull;

@Service
@Validated
public class DataAccessConfigService {

  private final DataAccessConfigRepository dataAccessConfigRepository;

  private final DataAccessFormService dataAccessFormService;

  private final DataAccessPreliminaryFormService dataAccessPreliminaryFormService;

  private final DataAccessFeasibilityFormService dataAccessFeasibilityFormService;

  private final DataAccessAmendmentFormService dataAccessAmendmentFormService;

  private final EventBus eventBus;

  @Inject
  public DataAccessConfigService(
    DataAccessConfigRepository dataAccessConfigRepository,
    DataAccessFormService dataAccessFormService,
    DataAccessPreliminaryFormService dataAccessPreliminaryFormService,
    DataAccessFeasibilityFormService dataAccessFeasibilityFormService,
    DataAccessAmendmentFormService dataAccessAmendmentFormService,
    EventBus eventBus) {
    this.dataAccessConfigRepository = dataAccessConfigRepository;
    this.dataAccessFormService = dataAccessFormService;
    this.dataAccessPreliminaryFormService = dataAccessPreliminaryFormService;
    this.dataAccessFeasibilityFormService = dataAccessFeasibilityFormService;
    this.dataAccessAmendmentFormService = dataAccessAmendmentFormService;
    this.eventBus = eventBus;
  }

  public synchronized DataAccessConfig getOrCreateConfig() {
    if (dataAccessConfigRepository.count() == 0) {
      DataAccessConfig config = new DataAccessConfig();
      dataAccessConfigRepository.save(config);
    }
    DataAccessConfig config = dataAccessConfigRepository.findAll().get(0);
    boolean modified = false;
    if (Strings.isNullOrEmpty(config.getCsvExportFormat())) {
      config.setCsvExportFormat(dataAccessFormService.getDefaultDataAccessFormResourceAsString("export-csv-schema.json"));
      modified = true;
    }
    if (Strings.isNullOrEmpty(config.getPreliminaryCsvExportFormat())) {
      config.setPreliminaryCsvExportFormat(dataAccessPreliminaryFormService.getDefaultDataAccessFormResourceAsString("export-csv-schema.json"));
      modified = true;
    }
    if (Strings.isNullOrEmpty(config.getFeasibilityCsvExportFormat())) {
      config.setFeasibilityCsvExportFormat(dataAccessFeasibilityFormService.getDefaultDataAccessFormResourceAsString("export-csv-schema.json"));
      modified = true;
    }
    if (Strings.isNullOrEmpty(config.getAmendmentCsvExportFormat())) {
      config.setAmendmentCsvExportFormat(dataAccessAmendmentFormService.getDefaultDataAccessFormResourceAsString("export-csv-schema.json"));
      modified = true;
    }
    if (modified) {
      dataAccessConfigRepository.save(config);
    }
    return config;
  }

  public void save(@NotNull @Valid DataAccessConfig config) {
    DataAccessConfig savedConfig = getOrCreateConfig();

    BeanUtils.copyProperties(config, savedConfig, "id", "version", "createdBy", "createdDate", "lastModifiedBy",
      "lastModifiedDate");

    dataAccessConfigRepository.save(savedConfig);
    eventBus.post(new DataAccessConfigUpdatedEvent(savedConfig));
  }

}
